### PR TITLE
Null Safety

### PR DESCRIPTION
**Pull request by micimize**
Supersedes #193 as the nullsafety tracking PR. Will remain a draft and reference point for prereleases until dart has a stable null safe release. @klavs am planning on publishing `*-nullsafety.0` versions for migrated packages soon, so that downstream packages can begin migrating.

Null safety migrations run thus far:

### Libraries
 - [X] gql:                [`0.13.0-nullsafety.1`](https://pub.dev/packages/gql/versions/0.13.0-nullsafety.1)
 - [X] gql_exec:           [`0.3.0-nullsafety.1`](https://pub.dev/packages/gql_exec/versions/0.3.0-nullsafety.1)
 - [X] gql_link:           [`0.4.0-nullsafety.1`](https://pub.dev/packages/gql_link/versions/0.4.0-nullsafety.1)
 - [x] gql_http_link:      [`0.4.0-nullsafety.1`](https://pub.dev/packages/gql_http_link/versions/0.4.0-nullsafety.1)
 - [x] gql_dedupe_link:    [`2.0.0-nullsafety.1`](https://pub.dev/packages/gql_dedupe_link/versions/2.0.0-nullsafety.1)
 - [x] gql_error_link:     [`0.2.0-nullsafety.1`](https://pub.dev/packages/gql_error_link/versions/0.2.0-nullsafety.1)
 - [x] gql_transform_link: [`0.2.0-nullsafety.1`](https://pub.dev/packages/gql_transform_link/versions/0.2.0-nullsafety.1)
 - [x] gql_websocket_link: [`0.3.0-nullsafety.1`](https://pub.dev/packages/gql_websocket_link/versions/0.3.0-nullsafety.1)
 - [x] gql_dio_link: [`0.1.0-nullsafety.1`](https://pub.dev/packages/gql_dio_link/versions/0.1.0-nullsafety.1)

#### ~~blocked by https://github.com/dart-lang/build/issues/2920~~
 - [x] gql_code_builder
 - [x] gql_build

### Examples
 - [X] cats
 - [x] gql_example_cli
 - [x] gql_example_cli_github
 - [x] gql_example_build
 - [x] gql_example_http_auth_link
 - [x] gql_example_dio_link
- [x] gql_example_flutter

### stable release TODOs
- [x] ~~Hand-verify `gql` ASTs against the [current spec](https://spec.graphql.org/June2018/)~~

### CODE REVIEW

**Comment by klavs** [comment]
Sorry, I reviewed part of the closed PR https://github.com/gql-dart/gql/pull/193

**Comment by micimize** [comment]
@klavs no prob – will take a look there

**Comment by klavs** [comment]
>  @klavs am planning on publishing *-nullsafety.0 versions for migrated packages soon, so that downstream packages can begin migrating.

I feel like I did not explicitly address this. Yes, please publish as you see fit.

File: gql/lib/src/language/lexer.dart [hunk 1]
```diff
@@ -162,15 +162,15 @@ class _Scanner {
     }
     consumeWhitespace();
 
-    if (position >= src.length) {
+    final code = peek();
+
+    if (code == null) {
```
Comment (micimize) [inline gql/lib/src/language/lexer.dart:167]: @klavs this is based on your earlier comment and the change in the initial draft. Though it worth noting as it removes the `position >= src.length` check entirely

File: gql/lib/src/language/parser.dart [hunk 2]
```diff
@@ -315,13 +315,13 @@ class _Parser {
 
   ValueNode _parseNonConstValue() => _parseValue(isConst: false);
 
-  ValueNode _parseValue({bool isConst}) {
-    final token = _next();
+  ValueNode _parseValue({bool? isConst}) {
+    final token = _next()!;
```
Comment (micimize) [inline gql/lib/src/language/parser.dart:319]: @klavs applied [the previous changes to `parser.dart`](https://github.com/gql-dart/gql/pull/181/files#diff-8eae12e33d2d5ab4ad734534b186e82d6f7e0a43684712daa2a6cd3a6c4b38f1), but a lot of these I don't really understand the expected behavior of yet

**Review by micimize — COMMENTED** [review]
Applied changes from #193 and incorporated some of the parser changes from #181

**Comment by agent3bood** [comment]
When running `dart pub outdated --mode=null-safety` I get 

`gql_pedantic  -        ✗1.0.2      ✗1.0.2      ✗1.0.3-alpha+1615272573346`


but  when trying to migrate `gql_pedantic` the migration tool says it is already support null-safety.

Did anybody had the same issue?

**Comment by andrewescutia** [comment]
excitedly awaiting the gql_dio_link update :-)

**Comment by agent3bood** [comment]
@micimize https://github.com/gql-dart/gql/pull/203 can be merged.

**Comment by micimize** [comment]
Given that the only packages remaining are examples, I think we can go ahead and merge this and publish stable version of the package-packages.

CI failures:
* check_svg – graphviz seems to generate different output between versions or something so there should just be a check for the gv input itself or something
* publish dry run – package:code_builder is not opted into null safety in its pubspec.yaml so it warns

**Comment by agent3bood** [comment]
@micimize I have opened PRs for migrating some examples. You can go ahead and merge it.

**Comment by micimize** [comment]
@agent3bood please make sure the CI for each example is succeeding in its PR first – I had to go fix tests for `gql_dio_link`.

You can add `fail-fast: false` to the CI for the examples to verify:
 https://github.com/gql-dart/gql/blob/2175bd080950de5340899096ed1cd934e2084fb5/.github/workflows/dart.yml#L7

**Comment by vasilich6107** [comment]
Hi. Could you update which recently got null-safety support
https://pub.dev/packages/code_builder

**Comment by vasilich6107** [comment]
For those who blocked I released https://pub.dev/packages/gql_code_builder_temp

**Comment by micimize** [comment]
finally got dry run working. Skipping svg as it's a pain

**Pull request merged**